### PR TITLE
fix: add filename length safety check with random suffix

### PR DESCRIPTION
**Pull request by Baraff24**
## Description
This pull request enhances the file upload process by introducing a new helper function, _ensure_safe_length, which ensures that filenames do not exceed the maximum allowed length (255 characters by default). If a filename is too long, it is truncated and a random hexadecimal suffix is appended to maintain uniqueness and comply with database constraints.

Details
Safe Filename Truncation:
The new _ensure_safe_length function checks the length of the filename and, when necessary, truncates it by reserving space for a random suffix. This suffix (default length of 16 characters) is generated using uuid.uuid4().hex and appended to the truncated filename.

Integration with get_valid_filename:
The get_valid_filename function has been updated to utilize _ensure_safe_length. In addition to slugifying the filename (to remove umlauts and special characters), it now guarantees that the resulting filename complies with the maximum length requirement.

Robustness in Upload Handling:
The changes further strengthen error handling during file uploads by ensuring that invalid content lengths or mismatches between MIME types and file extensions trigger a clear UploadException.

Reference:
This update addresses [Issue #1270](https://github.com/django-cms/django-filer/issues/1270), ensuring that filenames generated during uploads remain safe for storage.

## Related resources

issue: https://github.com/django-cms/django-filer/issues/1270

## Checklist

* [X] I have opened this pull request against ``master``
* [X] I have added or modified the tests when changing logic
* [X] I have followed [the conventional commits guidelines](https://www.conventionalcommits.org/) to add meaningful information into the changelog
* [X] I have read the [contribution guidelines ](https://github.com/django-cms/django-filer/blob/master/docs/development.rst#contributing) and I have joined #workgroup-pr-review on 
[Slack](https://www.django-cms.org/slack) to find a “pr review buddy” who is going to review my pull request.

## Summary by Sourcery

Implements a filename length safety check to ensure filenames do not exceed the maximum allowed length. It truncates the filename and appends a random hexadecimal suffix to maintain uniqueness and comply with database constraints.

Bug Fixes:
- Fixes issue where long filenames could cause errors during file uploads by truncating them and adding a random suffix.
- Ensures filenames comply with the maximum length requirement by integrating _ensure_safe_length into get_valid_filename.
- Strengthens error handling during file uploads by ensuring that invalid content lengths or MIME type mismatches trigger a clear UploadException.
- Addresses issue #1270, ensuring that filenames generated during uploads remain safe for storage.

### CODE REVIEW

**Review by fsbraun — COMMENTED** [review]
This looks good. I've added one comment to the tests. Also, can you go through the sourcery-ai comments on the tests. I think they make sense!

File: tests/test_files.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                          f"Unexpected character '{char}' found in filename.")
+
+    def test_edge_case_exact_length(self):
+        """
+        Test an edge case where the filename is exactly the maximum allowed length.
+        The function should leave such a filename unchanged.
+        """
+        # Create a filename that is exactly 255 characters long.
+        base = "b" * 250  # 250 characters for base
+        original = f"{base}.png"  # This may reach exactly or slightly above 255 depending on slugification
```
[the remark this answers is not in the document]
Comment (Baraff24) [reply tests/test_files.py:67]: Yes, you are right. I changed it to `base = "b" * 251`

**Comment by Baraff24** [comment]
> This looks good. I've added one comment to the tests. Also, can you go through the sourcery-ai comments on the tests. I think they make sense!

I have integrated all the tests suggested by sourcery-ai. Tell me if you think I should change anything else, thanks :)

**Comment by fsbraun** [comment]
@Baraff24 I came across another related issue: #1093 

The connection is this: Filer generates image thumbnails, these thumbnails have a fixed(see #1093) name being the original name plus some scaling info plus the original suffix. **We need to account for space the the thumbnail scaling info.** Maybe you can figure out what the longest possible scaling info is, otherwise I suggest to at least keep space for, say, 100 extra characters for the thumbnails.

**Comment by Baraff24** [comment]
> @Baraff24 I came across another related issue: #1093
> 
> The connection is this: Filer generates image thumbnails, these thumbnails have a fixed(see #1093) name being the original name plus some scaling info plus the original suffix. **We need to account for space the the thumbnail scaling info.** Maybe you can figure out what the longest possible scaling info is, otherwise I suggest to at least keep space for, say, 100 extra characters for the thumbnails.

@fsbraun The first solution that came to mind was to truncate to 155 characters with the same system in order to leave room for any other characters for all the image sizing settings. What do you think?

**Comment by fsbraun** [comment]
@Baraff24 Yes, I think that's necessary!

**Comment by Baraff24** [comment]
> @Baraff24 Yes, I think that's necessary!

@fsbraun in the last commit, I added two variables in the settings. Is this enough or do you want to set me up the variables in a static way?

**Comment by fsbraun** [comment]
Honestly, I prefer fixed values in the code (some constant). But it would need to be correct, e.g., 155, but not 255.

**Comment by Baraff24** [comment]
> Honestly, I prefer fixed values in the code (some constant). But it would need to be correct, e.g., 155, but not 255.

@fsbraun I applied your advices into my code. Check it now if you want

File: filer/utils/files.py [hunk 2]
```diff
@@ -121,6 +123,33 @@ def slugify(string):
     return slugify_django(force_str(string))
 
 
+def _ensure_safe_length(filename, max_length=155, random_suffix_length=16):
+    """
+    Ensures that the filename does not exceed the maximum allowed length.
+    If it does, the function truncates the filename and appends a random hexadecimal
+    suffix of length `random_suffix_length` to ensure uniqueness and compliance with
+    database constraints.
```
Comment (fsbraun) [inline filer/utils/files.py:131]:
```suggestion
    database constraints - even after markers for a thumbnail are added.
```

**Review by fsbraun — APPROVED** [review]
@Baraff24 Thanks! Great work!

File: filer/utils/files.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
 import mimetypes
 import os
+import uuid
 
+from django.conf import settings
```
Comment (fsbraun) [inline filer/utils/files.py:5]:
```suggestion
```

File: tests/test_files.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
+import string
+
+from django.conf import settings
```
Comment (fsbraun) [inline tests/test_files.py:3]:
```suggestion
```

**Review by fsbraun — COMMENTED** [review]
Only thing left: Remove the unused imports... ;-)

**Comment by Baraff24** [comment]
> Only thing left: Remove the unused imports... ;-)

@fsbraun already done. Now can we merge?
